README.md
[![published](https://static.production.devnetcloud.com/codeexchange/assets/images/devnet-published.svg)](https://developer.cisco.com/codeexchange/github/repo/CiscoDevNet/homebridge-meraki-mr)
[![npm](https://badgen.net/npm/v/homebridge-meraki-mr)](https://www.npmjs.com/package/homebridge-meraki-mr)[![npm](https://badgen.net/npm/dw/homebridge-meraki-mr)](https://www.npmjs.com/package/homebridge-meraki-mr)[![npm](https://badgen.net/npm/dt/homebridge-meraki-mr)](https://www.npmjs.com/package/homebridge-meraki-mr)

# A [Cisco Meraki](https://meraki.cisco.com) Wireless plugin for [Homebridge](https://github.com/nfarina/homebridge).  

---

# Purpose

The purpose of this code is to create a Homebridge plugin to allow a wireless administrator to enable or disable a WLAN through a Light Switch control in Apple’s homekit or through Siri.

![alt text](mrexample.png "Dashboard Example Screenshot")

![alt text](mrexample2.png "WLANs Example Screenshot")

# Installation
Run these commands:

    % sudo npm install -g homebridge
    % sudo npm install -g homebridge-meraki-mr


NOTE: If you install homebridge like this:

    sudo npm install -g --unsafe-perm homebridge

Then all subsequent installations must be like this:

    sudo npm install -g --unsafe-perm homebridge-meraki-mr

# Configuration
Example accessory config (needs to be added to the homebridge config.json):
 ...

	"platforms": [
		{
			"name": "Meraki WiFi",
			"apiKey": "4854372bf7b61bdb701512314e3134bcb3e61b85",
			"timeout": 10000,
			"refreshInterval": 15,
			"debug": false;
			"networks": [
				{
					"displayName": "Site 1",
					"networkId": "L_012345678901234501"
				},
				{
					"displayName": "Site 2",
					"networkId": "L_012345678901234502"
				}
			],
			"platform": "Meraki MR"
		}
	]
 ...

### Config Explanation:

Field           						| Description
----------------------------|------------
**platform**	   						| (required) Must always be "Meraki MR".
**name**										| (required) The internal name you want to use for the platform within Homebridge.
**apiKey**									| (required) The Meraki API Key used to access the Meraki Dashboard.
**timeout**									| (optional) The timeout duration in ms for the web API calls.
**refreshInterval**					| (optional) The refresh interval in minutes for polling the WiFi status.
**debug**										| (optional) Enables additional logging.
**displayName**							| (required) The name you want to show for the WLC in homekit.
**networkId**								| (required) The Meraki Network ID for the site or network.

To make your Meraki Dashboard account work with the plugin:

1. Login to your Meraki Dashboard account.
2. In the Dashboard, go to Organization / Settings.
3. Scroll down to the bottom and Enable access to the Cisco Meraki Dashboard API.
4. Save Changes, then scroll to the top, click on your username on the right & click My Profile.
5. Scroll down towards the bottom and under API Access, click on Generate API Key.
6. Write down or copy the API Key.
7. In the same browser, go to https://api.meraki.com/api/v0/organizations
8. Use the resulting org id to browse to https://https://api.meraki.com/api/v0/organizations/{orgID}/networks
9. Write down the network id(s) for the network(s) that you want to control.
10. Create your config file according to the above example (or using the Homebridge UI).


# More information
Check out https://github.com/nfarina/homebridge for more information about Homebridge.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.1.1 (2023-12-05)

#### Changes

- Updated API request in platform accessory.

## 0.1.0 (2023-12-05)

#### Changes

- Updated dependencies.  Updated API request.

## 0.0.3 (2021-12-31)

#### Changes

- Updated dependencies.

## 0.0.2 (2021-11-08)

#### Changes

- Moved WiFi status updates to polling method.

## 0.0.1 (2021-11-08)

#### Changes

- Initial commit.


mrplatformAccessory.js
'use strict'
const baseUrlV1 = "https://api.meraki.com/api/v1"

/**
 * Platform Accessory
 * An instance of this class is created for each accessory your platform registers
 * Each accessory may expose multiple services of different service types.
 */
class MRPlatformAccessory {
//private service: Service;

  constructor(platform,accessory){
  	this.platform = platform
  	this.accessory = accessory
  	  	
    for (let i = 0; i < this.accessory.context.device.wlanCount; i++) {
	    const iString = i.toString()
	    
	    const switchService = this.accessory.getService(iString) ?? this.accessory.addService(this.platform.Service.Switch, this.accessory.context.device.ssidList[i].ssid, iString)
      // To avoid "Cannot add a Service with the same UUID another Service without also defining a unique 'subtype' property." error,
      // when creating multiple services of the same type, you need to use the following syntax to specify a name and subtype id:
      // this.accessory.getService('NAME') ?? this.accessory.addService(this.platform.Service.Switch, 'NAME', 'USER_DEFINED_SUBTYPE');

      // register handlers for the On/Off Characteristic
      switchService.getCharacteristic(this.platform.Characteristic.On)
        .on('set', this.setPowerState.bind(this, i))   // SET - bind to the `setPowerState` method below
//        .on('get', this.getPowerState.bind(this, i))		// GET - bind to the `getPowerState` method below
    }
 
    // set accessory information
    
    this.accessory.getService(this.platform.Service.AccessoryInformation)
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Meraki')
      .setCharacteristic(this.platform.Characteristic.Model, this.accessory.context.device.model)
      .setCharacteristic(this.platform.Characteristic.SerialNumber, this.accessory.context.device.serial)
      .setCharacteristic(this.platform.Characteristic.FirmwareRevision, this.accessory.context.device.firmware)
		this.timer = setTimeout(this.poll.bind(this), this.platform.refreshInterval)
		this.poll()
	}
/*
	async getPowerState(index, callback) {
		try {
			const response = await this.platform.session({
				method: 'get',
				url: baseUrlV1 + "/networks/" + this.accessory.context.device.networkId + "/ssids/" + this.accessory.context.device.ssidList[index].wlanNumber,
				data: {},
				headers: {
					"X-Cisco-Meraki-API-Key": this.platform.apiKey,
					"Accept": "application/json",
					"Content-Type": "application/json"
				},
				timeout: this.platform.timeout
			}).catch(err => {
					this.platform.log.error('Error getting WLAN state %s',err)
			})
			var powerState = response.data.enabled
			this.accessory.context.device.ssidList[index].state = powerState
			this.platform.log('Get WLAN state for %s = %s',  this.accessory.context.device.ssidList[index].ssid, powerState)
			callback (null, powerState)
		}catch(err) {
			this.platform.log.error('Error getting WLAN state %s', err)
			callback (err, null)
		}
	}
*/
	async setPowerState(index, powerState, callback) {
		try {
//			this.platform.log('Setting WLAN state for %s = %s', this.accessory.context.device.ssidList[index].ssid, powerState)
			const response = await this.platform.session({
				method: 'put',
				url: baseUrlV1 + "/networks/" + this.accessory.context.device.networkId + "/wireless/ssids/" +  this.accessory.context.device.ssidList[index].wlanNumber,
				data: { 'enabled': powerState},
				headers: {
					"X-Cisco-Meraki-API-Key": this.platform.apiKey,
					"Accept": "application/json",
					"Content-Type": "application/json"
				},
				timeout: this.platform.timeout
			}).catch(err => {
					this.platform.log.error('Error setting WLAN state %s',err)
					this.platform.log.warn(err.response.data.errors)
			})
			this.accessory.context.device.ssidList[index].state = powerState
			this.platform.log("Set " + this.accessory.context.device.ssidList[index].ssid + " to " + powerState)
			this.updateUI(index)
			callback (null)
		}catch(err) {
			this.platform.log.error('Error setting WLAN state %s', err)
			callback (err)
		}
	}

	//
	//	This will update the homekit UI

	async updateUI(index) {
		setTimeout( () => {
			this.accessory.services[index+1].getCharacteristic(Characteristic.On).updateValue(this.accessory.context.device.ssidList[index].state)
			if (this.platform.debug) this.platform.log(this.accessory.context.device.ssidList[index].ssid + " = " + this.accessory.context.device.ssidList[index].state)
		}, 100)
	}
	
	async poll() {
		if(this.timer) clearTimeout(this.timer)
		this.timer = null
		try {
			const response = await this.platform.session({
				method: 'get',
				url: baseUrlV1 + "/networks/" + this.accessory.context.device.networkId + "/wireless/ssids",
				data: {},
				headers: {
					"X-Cisco-Meraki-API-Key": this.platform.apiKey,
					"Accept": "application/json",
					"Content-Type": "application/json"
				},
				timeout: this.platform.timeout
			}).catch(err => {
					this.platform.log.error('Error getting WLAN state %s',err)
			})
			for (let i = 0; i < this.accessory.context.device.wlanCount; i++) {
				let obj = response.data.find(o => o.number === this.accessory.context.device.ssidList[i].wlanNumber)
				this.accessory.context.device.ssidList[i].state = obj.enabled				
				this.updateUI(i)
			}
		}catch(err) {
				this.platform.log.error('Error getting WLAN state %s',err)
		}
		this.timer = setTimeout(this.poll.bind(this), this.platform.refreshInterval)
	}

}

module.exports=MRPlatformAccessory;

mrplatform.js
'use strict'

//
//	We are going to use axios with SSL for our RESTconf calls to the WLC,
//	so we'll need these node-modules

const https = require('https')
const axios = require('axios')
axios.defaults.withCredentials = true

//
//	Setting up a few constant variables for reference

const packageJson=require('./package')
const MRPlatformAccessory=require('./mrplatformAccessory')

const baseUrlV1 = "https://api.meraki.com/api/v1"
const timeout = 10000
const interval = 15 // Minutes
const debug = false
const	PluginName = packageJson.name
const version = packageJson.version

class MRPlatform {

	//
	//	Setup our instance variables based on the homebridge config.json entries

	constructor(log, config, api) {

		this.log = log
		this.config = config
		this.apiKey = config["apiKey"]
		this.networkId = config["networkId"]
		this.timeout = config["timeout"] === undefined ? timeout : config["timeout"]
		this.refreshInterval = config["refreshInterval"] === undefined ? (interval * 60000) : (config["refreshInterval"] * 60000)
		this.debug = config["debug"] === undefined ? debug : config["debug"]
		this.accessories = []
		this.session = axios.create({
			timeout: this.timeout,
			httpsAgent: new https.Agent({  
				rejectUnauthorized: false
				}),
		})

		if(api){
			this.api=api
			this.Service = this.api.hap.Service
  		this.Characteristic = this.api.hap.Characteristic
			this.api.on("didFinishLaunching", function(){
				//Get devices
				this.getMRDevices()
			}.bind(this))
		}
	}
	
	identify(){
		this.log('Identify the WLC')
	}
	
	configureAccessory(accessory) {
    if (this.config.networks.find( (network) => network.id === accessory.context.device.id ))  {
			if (!this.accessories.find( (anAccessory) => anAccessory.context.device.id === accessory.context.device.id ))  {
				this.log.info('Restoring accessory from cache:', accessory.displayName)

				// create the accessory handler
				// this is imported from `c9000platformAccessory.js`
				new MRPlatformAccessory(this, accessory)

				// add the restored accessory to the accessories cache so we can track if it has already been registered
				this.accessories.push(accessory)
			}
    }
	}
	
	getMRDevices(){

		//
    // Check for a blank config and return without registering accessories

		if(!this.config){
			this.log.warn('Ignoring MR Platform setup because it is not configured')
			return
		}
		
    // loop over the discovered devices and register each one if it has not already been registered
		for (const device of this.config.networks){
			this.log.debug('Getting network list')
			var networkDevice = {
				"displayName": device.displayName,
				"networkId": device.networkId,
				"model": "MR",
				"serial": "dummy-serial",
				"firmware": version,
				"wlanCount": 0,
				"ssidList": [],
			}
			const webConfig = {
				data: {},
				headers: {
					"X-Cisco-Meraki-API-Key": this.apiKey,
					"Accept": "application/json",
					"Content-Type": "application/json"
				}
			}

			this.session.get(baseUrlV1 + "/networks/" + networkDevice.networkId + "/wireless/ssids", webConfig)
				.then(resp => {
					if (this.debug) {
						this.log(baseUrlV1 + "/networks/" + networkDevice.networkId + "/wireless/ssids");
					}
					networkDevice.wlanCount = Object.keys(resp.data).length;
					this.log("Dashboard response: " + networkDevice.wlanCount + " SSID(s).");
			
					var ssids = resp.data;

					for (var i = 0; i < networkDevice.wlanCount; i++) {
						if (ssids[i].name.search("Unconfigured SSID ") != 0)
						{
							if (this.debug) {
								this.log(i, ssids[i].name + " = " + ssids[i].number);
							}
							var wlanNumber = ssids[i].number
							var ssid = ssids[i].name
							var state = ssids[i].enabled
							networkDevice.ssidList.push({wlanNumber,ssid,state})
						}
					}
					networkDevice.wlanCount = networkDevice.ssidList.length
					if (this.debug) this.log(networkDevice.ssidList)
					
					// generate a unique id for the accessory this should be generated from
					// something globally unique, but constant, for example, the device serial
					// number or MAC address
					const uuid = this.api.hap.uuid.generate(device.networkId);

					// check that the device has not already been registered by checking the
					// cached devices we stored in the `configureAccessory` method above
					if (!this.accessories.find(accessory => accessory.UUID === uuid)) {
						this.log.info('Registering new accessory:', device.displayName, uuid)

						// create a new accessory
						const accessory = new this.api.platformAccessory(device.displayName, uuid)

						// store a copy of the device object in the `accessory.context`
						// the `context` property can be used to store any data about the accessory you may need
						accessory.context.device = networkDevice
	
						// create the accessory handler
						// this is imported from `platformAccessory.js`
						new MRPlatformAccessory(this, accessory)

						// link the accessory to your platform
						this.api.registerPlatformAccessories(PluginName, PlatformName, [accessory])

						// push into accessory cache
						this.accessories.push(accessory)
						this.log.info('New accessory:', accessory.displayName)

						// it is possible to remove platform accessories at any time using `api.unregisterPlatformAccessories`, eg.:
						// this.api.unregisterPlatformAccessories(PluginName, PlatformName, [accessory]);
					}
				})
				.catch(err => {
					this.log.error(err.stack)
				})
		}
	}
}

module.exports=MRPlatform;